[cmd3] Add rising and falling edge trigger factories

SamCarlberg opened the pull request:
`Trigger.getAsBoolean()` behavior has been changed from passing through the underlying boolean supplier to returning the latest cached signal as determined by the most recent call to `poll()`. This allows rising and falling edge triggers to have a consistent return value over an entire polling cycle, rather than only being high for the _first_ check in a cycle.

Closes #8309

SamCarlberg committed:
Add rising and falling edge triggers

SamCarlberg committed:
Bind triggers to event loops upon construction

Only binding when a command binding is added would have resulted in intermediate edge triggers never firing

KangarooKoala commented:
When [we made `EventLoop` bind on construction](https://github.com/wpilibsuite/allwpilib/pull/5436), it [caused issues](https://github.com/wpilibsuite/allwpilib/issues/5903) since then the instances would never be GC'd. We then added button caching to fix the issue, so it's not so much of an issue, but I'm still hesitant to introduce that kind of behavior.

SamCarlberg committed:
Bind to event loop on binding; track dependencies recursively

This allows intermediate unbound triggers to still function

KangarooKoala reviewed (commented):
Generally seems okay to me (someone who has not really interacted with commandsv3, just commandsv2 and BooleanEvent), for what it's worth. Hopefully someone else could also review?

SamCarlberg committed:
Merge branch '2027' into commandsv3/trigger-edges

SamCarlberg committed:
Eagerly bind triggers to the event loop

Means that we don't need to track dependency triggers. However, this means triggers can never be garbage collected because they're always referenced by their event loops (unless users manually clear the event loop, which is unlikely)

SamCarlberg commented on commandsv3/src/main/java/org/wpilib/command3/Trigger.java:
```diff
@@ -177,9 +204,21 @@ public Trigger toggleOnFalse(Command command) {
+   * from {@code getAsBoolean()} may not agree with the result of checking the signal directly.
+   *
+   * <p>Triggers are only bound to an event loop when at least one command has been bound to the
+   * trigger. This method will always return {@code false} before commands have been bound.
```
I've changed the implementation to eagerly bind triggers to event loops. Otherwise there's too much opportunity for footguns when checking triggers that may or may not be bound to the event loop.

The downside is that this opens up risks for memory leaks: triggers can't be GC'd unless users manually clear the scheduler's event loop.

SamCarlberg committed:
Merge remote-tracking branch 'wpi/2027' into commandsv3/trigger-edges

SamCarlberg committed:
Track trigger creation scopes and automatically clean up when they exit

Simliar to command binding scopes, triggers will be unbound from their event loops when their creation scopes become inactive

This helps prevent issues where triggers are created in opmodes or in commands and would never be cleaned up, causing potential performance problem (by polling triggers that are not bound and will never get new bindings) or memory leaks (due to tracking these zombie triggers in lists with hard references)

Daniel1464 commented:
It would be useful to add a javac plugin that makes using whileTrue/whileFalse with rising edge/falling edge triggers a compile-time error.
